d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -1447,6 +1462,10 @@ private void handleRpcStreamClosed(Status error)
         delayNanos = _retryBackoffPolicy.nextBackoffNanos();
       }
       _log.info("Retry ADS stream in {} ns", delayNanos);
+      if (_shutdown) {
+        _log.warn("Already shut down, will not retry ADS stream");
+        return;
+      }
       _retryRpcStreamFuture = _executorService.schedule(new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS);
```

Comment:
nit: we should also have a helper function checkAndSchedule (or checkShutdownAndSchedule), and call it here. There are other usage of .schedule in startRpcStreamLocal and testRetryTask. When it won't hurt to guard them, People unfamiliar with the code shouldn't need to look into the code and wonder why they don't need to be guarded, we should just guard all usages to be 100% safe.

Reply:
Makes sense, it's just a little extra code to make sure we are protecting every surface area

Reply:
One challenge with `checkShutdownAndSchedule` is that `_executorService.schedule` returns a `ScheduledFuture<V>` and it's not clear what we should return when the `_executorService` is shutdown.

I have 
```
  private ScheduledFuture<?> checkShutdownAndSchedule(Runnable runnable, int delay, TimeUnit unit) {
    if (_executorService.isShutdown())
    {
      _log.warn("Attempting to schedule a task after _executorService was shutdown, will do nothing");
      return null;
    }

    return _executorService.schedule(runnable, delay, unit);
  }
```

but I worry that passing null to the caller may lead to a `NullPointerException` somewhere

Reply:
Update: I checked all the places `checkShutdownAndSchedule` is called and the variable that the result is assigned to is always checked for null, so we should actually be good here